File: app/src/main/java/com/romanov/labs/lab4/MainActivity.java
package com.romanov.labs.lab4;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;

import androidx.activity.result.ActivityResult;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.romanov.labs.lab4.databinding.ActivityMainBinding;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {
    public static final String EXTRA_CARD = "extra_card";
    private ActivityMainBinding binding = null;
    private CardAdapter adapter = null;
    ArrayList<Card> cards = new ArrayList<>();
    private ItemTouchHelper itemTouchHelper;

    @SuppressLint("NotifyDataSetChanged")
    private void onAddCardResult(ActivityResult result) {
        if (result.getResultCode() == RESULT_OK) {
            Intent data = result.getData();
            if (data != null) {
                Card addedCard = (Card) data.getSerializableExtra(EXTRA_CARD);
                if (addedCard != null) {
                    cards.add(addedCard);
                    XmlHelper.saveToXml(MainActivity.this, cards);
                    notifyDataSetChanged();
                }
            }
        }
    }

    private final ActivityResultLauncher<Intent> addCardLauncher = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), this::onAddCardResult);

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        LinearLayoutManager manager = new LinearLayoutManager(this);
        binding.recyclerView.setLayoutManager(manager); // setInitialData();
        cards.addAll(XmlHelper.readFromXml(MainActivity.this, "cards.xml"));
        notifyDataSetChanged();
        ItemTouchHelper.SimpleCallback simpleCallback = new ItemTouchHelper.SimpleCallback(
                0, ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT) {
            @Override
            public boolean onMove(@NonNull RecyclerView recyclerView,
                                  @NonNull RecyclerView.ViewHolder viewHolder,
                                  @NonNull RecyclerView.ViewHolder target) {
                // Not implementing as we're not considering item movement
                return false;
            }

            @Override
            public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
                // Handling swipe (deleting item)
                int position = viewHolder.getAdapterPosition();
                showDeleteConfirmationDialog(position);
            }
        };

        itemTouchHelper = new ItemTouchHelper(simpleCallback);
        itemTouchHelper.attachToRecyclerView(binding.recyclerView);
    }

    private void notifyDataSetChanged() { // Обработка свайпа (удаление элемента)
        adapter = new CardAdapter(MainActivity.this, cards, null); // создаем адаптер
        binding.recyclerView.setAdapter(adapter); // устанавливаем для списка адаптер
    }

    private void showDeleteConfirmationDialog(int position) {
        binding.recyclerView.setAdapter(adapter); // устанавливаем для списка адаптер
        AlertDialog.Builder builder = new AlertDialog.Builder(MainActivity.this);
        builder.setTitle("Подтверждение удаления");
        builder.setMessage("Вы уверены, что хотите удалить эту карточку?");
        builder.setPositiveButton("Да", (dialog, which) -> {
            adapter.removeItem(position);
            cards.remove(position);
            XmlHelper.saveToXml(MainActivity.this, cards);
            binding.recyclerView.setAdapter(adapter); // устанавливаем для списка адаптер
//            deleteListener.onItemDelete(position);
        });
        builder.setNegativeButton("Нет", (dialog, which) -> {
            dialog.dismiss();
        });
        builder.show();
    }

    public void addCard(View view) {
        Intent intent = new Intent(this, EditCardActivity.class);
        addCardLauncher.launch(intent);
    }
}

File: app/src/main/java/com/example/myApp/MainActivity.kt
package com.example.myApp

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import android.app.AlertDialog
import com.example.myApp.Data.CallbackFun
import com.example.myApp.Data.CardsAdapter
import com.example.myApp.Data.CardsRepository
import com.example.myApp.Data.TermCard
import com.example.myApp.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var adapter: CardsAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupRecyclerView()
        setupSwipeToDelete()
        setupAddButton()
    }

    override fun onResume() {
        super.onResume()
        updateCards()
    }

    private fun setupRecyclerView() {
        adapter = CardsAdapter(adapterCallBack)
        binding.recyclerView.layoutManager = LinearLayoutManager(this)
        binding.recyclerView.adapter = adapter
        updateCards()
    }

    private fun setupAddButton() {
        binding.buttonAddItem.setOnClickListener {
            val intent = Intent(this, AddCardActivity::class.java)
            startActivity(intent)
        }
    }

    private fun updateCards() {
        adapter.setItem(CardsRepository.getCards())
        adapter.notifyDataSetChanged()
    }

    private val adapterCallBack = object : CallbackFun {
        override fun deleteCard(card: TermCard) {
            AlertDialog.Builder(this@MainActivity).apply {
                setTitle("Подтвердите удаление")
                setMessage("Вы уверены, что хотите удалить эту карточку?")
                setPositiveButton("Удалить") { dialog, which ->
                    CardsRepository.deleteCard(card)
                    adapter.notifyDataSetChanged() // Обновляем адаптер после удаления
                }
                setNegativeButton("Отмена", null)
            }.create().show()
        }

        override fun showCard(index: Int) {
            val intent = Intent(this@MainActivity, ViewCard::class.java)
            intent.putExtra("index", index)
            this@MainActivity.startActivity(intent)
        }
    }

    private fun setupSwipeToDelete() {
        val itemTouchHelper = ItemTouchHelper(object :
            ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT) {
            override fun onMove(
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder,
                target: RecyclerView.ViewHolder
            ): Boolean {
                return false
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                val position = viewHolder.adapterPosition
                val card = adapter.getItemAtPosition(position)
                adapterCallBack.deleteCard(card)
            }
        })

        itemTouchHelper.attachToRecyclerView(binding.recyclerView)
    }
}
